governance/program/src/state/proposal.rs

```diff
@@ -102,11 +110,182 @@ impl Proposal {
+    pub fn assert_can_vote(
+        &self,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        if self.state != ProposalState::Voting {
+            return Err(GovernanceError::InvalidStateCannotVote.into());
+        }
+
+        // Check if we are still within the configured max_voting_time period
+        if self
+            .voting_at
+            .unwrap()
+            .checked_add(config.max_voting_time)
+            .unwrap()
+            < current_slot
+        {
+            return Err(GovernanceError::ProposalVotingTimeExpired.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks if Proposal can be finalized
+    pub fn assert_can_finalize_vote(
+        &self,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        if self.state != ProposalState::Voting {
+            return Err(GovernanceError::InvalidStateCannotFinalize.into());
+        }
+
+        // Check if we passed the configured max_voting_time period yet
+        if self
+            .voting_at
+            .unwrap()
+            .checked_add(config.max_voting_time)
+            .unwrap()
+            >= current_slot
+        {
+            return Err(GovernanceError::CannotFinalizeVotingInProgress.into());
+        }
+
+        Ok(())
+    }
+
+    /// Finalizes vote by moving it to final state Succeeded or Defeated if max_voting_time has passed
+    /// If Proposal is still within max_voting_time period then error is returned
+    pub fn finalize_vote(
+        &mut self,
+        governing_token_supply: u64,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
+        self.assert_can_finalize_vote(config, current_slot)?;
+
+        let current_yes_vote_threshold =
+            self.get_current_vote_threshold(self.yes_votes_count, governing_token_supply);
+
+        let final_state = if current_yes_vote_threshold > config.vote_threshold_percentage as u64 {
+            ProposalState::Succeeded
+        } else {
+            ProposalState::Defeated
+        };
+
+        self.state = final_state;
+        self.voting_completed_at = Some(current_slot);
+
+        Ok(())
+    }
+
+    /// Returns current vote threshold in relation to the total governing token supply
+    fn get_current_vote_threshold(&self, vote_count: u64, governing_token_supply: u64) -> u64 {
+        vote_count
+            .checked_mul(100)
+            .unwrap()
+            .checked_div(governing_token_supply)
+            .unwrap()
+    }
+
+    /// Checks if vote can be tipped and automatically transitioned to Succeeded or Defeated state
+    /// If the conditions are met the state is updated accordingly
+    pub fn try_tip_vote(
+        &mut self,
+        governing_token_supply: u64,
+        config: &GovernanceConfig,
+        current_slot: Slot,
+    ) -> Result<(), ProgramError> {
```

Comment:
You are right, this should not return Result because the function only attempts to tip the vote if possible and has no error conditions. If anything it should be bool to indicate tipped/not tipped but since I never use the return value I removed it altogether.